tensorflow/contrib/s3/s3_file_system.cc

```diff
@@ -0,0 +1,578 @@
+static const size_t S3ReadAppendableFileBufferSize = 1024 * 1024;
+static const int S3GetChildrenMaxKeys = 100;
+
+Aws::Client::ClientConfiguration& GetDefaultClientConfig() {
+  static std::mutex cfg_lock;
+  static bool init(false);
+  static Aws::Client::ClientConfiguration cfg;
+
+  std::lock_guard<std::mutex> lock(cfg_lock);
+
+  if (!init) {
+    const char* endpoint = getenv("TFS3_ENDPOINT");
+    if (endpoint) {
+      cfg.endpointOverride = Aws::String(endpoint);
+    }
+    const char* region = getenv("TFS3_REGION");
+    if (region) {
+      cfg.region = Aws::String(region);
+    }
+    const char* use_https = getenv("TFS3_USE_HTTPS");
+    if (use_https) {
+      if (use_https[0] == '0') {
+        cfg.scheme = Aws::Http::Scheme::HTTP;
+      } else {
+        cfg.scheme = Aws::Http::Scheme::HTTPS;
+      }
+    }
+    const char* verify_ssl = getenv("TFS3_VERIFY_SSL");
+    if (verify_ssl) {
+      if (verify_ssl[0] == '0') {
+        cfg.verifySSL = false;
+      } else {
+        cfg.verifySSL = true;
+      }
+    }
+
+    init = true;
+  }
+
+  return cfg;
+};
+
+Status ParseS3Path(const string& fname, bool empty_object_ok, string* bucket,
+                   string* object) {
+  if (!bucket || !object) {
+    return errors::Internal("bucket and object cannot be null.");
+  }
+  StringPiece scheme, bucketp, objectp;
+  io::ParseURI(fname, &scheme, &bucketp, &objectp);
+  if (scheme != "s3") {
+    return errors::InvalidArgument("S3 path doesn't start with 's3://': ",
+                                   fname);
+  }
+  *bucket = bucketp.ToString();
+  if (bucket->empty() || *bucket == ".") {
+    return errors::InvalidArgument("S3 path doesn't contain a bucket name: ",
+                                   fname);
+  }
+  objectp.Consume("/");
+  *object = objectp.ToString();
+  if (!empty_object_ok && object->empty()) {
+    return errors::InvalidArgument("S3 path doesn't contain an object name: ",
+                                   fname);
+  }
+  return Status::OK();
+}
+
+class S3RandomAccessFile : public RandomAccessFile {
+ public:
+  S3RandomAccessFile(const string& bucket, const string& object)
+      : bucket_(bucket), object_(object) {}
+
+  Status Read(uint64 offset, size_t n, StringPiece* result,
+              char* scratch) const override {
+    Aws::S3::S3Client s3Client(GetDefaultClientConfig());
+    Aws::S3::Model::GetObjectRequest getObjectRequest;
+    getObjectRequest.WithBucket(bucket_.c_str()).WithKey(object_.c_str());
+    char buffer[50];
+    memset(buffer, 0x00, sizeof(buffer));
+    snprintf(buffer, sizeof(buffer) - 1, "bytes=%lld-%lld", offset,
+             offset + n - 1);
+    getObjectRequest.SetRange(buffer);
+    getObjectRequest.SetResponseStreamFactory([]() {
+      return Aws::New<Aws::StringStream>(S3FileSystemAllocationTag);
+    });
+    auto getObjectOutcome = s3Client.GetObject(getObjectRequest);
+    if (!getObjectOutcome.IsSuccess()) {
+      n = 0;
```

Comment:
Does this work for short reads? We still want to return the bytes read even with OUT_OF_RANGE.

For example, if we requested 10 bytes, but only read 5 of them, then we'll return with OUT_OF_RANGE but return the 5 bytes.